<!DOCTYPE html>
<html lang="en"><head>
<script src="index_files/libs/clipboard/clipboard.min.js"></script>
<script src="index_files/libs/quarto-html/tabby.min.js"></script>
<script src="index_files/libs/quarto-html/popper.min.js"></script>
<script src="index_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="index_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="index_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="index_files/libs/quarto-html/quarto-syntax-highlighting-7b4406b7675125bc2ba204020e191172.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.7.22">

  <meta name="author" content="Mattia S.">
  <title>on the positional BWT</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="index_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="index_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="index_files/libs/revealjs/dist/theme/quarto-00ac2f6af1389919a18f42df5b7c0002.css">
  <link href="index_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="index_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="index_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="index_files/libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="index_files/libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="index_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">on the positional BWT</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Mattia S. 
</div>
</div>
</div>

</section>
<section id="bf-pbwt-results" class="title-slide slide level1 center">
<h1>bf-pbwt results</h1>
<table class="caption-top">
<thead>
<tr class="header">
<th>BLIM timing(s)</th>
<th>chr10(18gb)</th>
<th>chr21(4.7gb)</th>
<th>chr22(4.8)</th>
<th></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>2bfpbwt-bm on master</td>
<td>690.53</td>
<td>110.47</td>
<td>116.99</td>
<td></td>
</tr>
<tr class="even">
<td>2bfpbwt on (71a045a)</td>
<td>294.98</td>
<td>74.48</td>
<td>75.54</td>
<td></td>
</tr>
<tr class="odd">
<td>2bfpbwt-bm on (hotfix)</td>
<td>753.71</td>
<td>108.56</td>
<td>113.40</td>
<td></td>
</tr>
</tbody>
</table>
<table class="caption-top">
<thead>
<tr class="header">
<th>BLIM ram (GiB)</th>
<th>chr10(18gb)</th>
<th>chr21(4.7gb)</th>
<th>chr22(4.8)</th>
<th></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>2bfpbwt-bm on master</td>
<td>250</td>
<td>82</td>
<td>84</td>
<td></td>
</tr>
<tr class="even">
<td>2bfpbwt on (71a045a)</td>
<td>158</td>
<td>43</td>
<td>44</td>
<td></td>
</tr>
<tr class="odd">
<td>2bfpbwt-bm on (hotfix)</td>
<td>244</td>
<td>82</td>
<td>84</td>
<td></td>
</tr>
</tbody>
</table>
</section>

<section>
<section id="review" class="title-slide slide level1 center">
<h1>REVIEW</h1>

</section>
<section id="basic-pbwt-durbinefficienthaplotypematching2014" class="slide level2">
<h2>Basic PBWT <span class="smaller"><span class="citation" data-cites="durbinEfficientHaplotypeMatching2014">[<a href="#/references" role="doc-biblioref" onclick="">1</a>]</span></span></h2>
<p><strong>Input</strong>:</p>
<p>a binary matrix <span class="math inline">\(X\)</span>, of <span class="math inline">\(m\)</span> rows and <span class="math inline">\(n\)</span> columns. <span class="math inline">\((\Sigma = \{0, 1\})\)</span>, where <span class="math inline">\(x_i\)</span> are haplotype sequences; <span class="math inline">\(0\leq i &lt; m\)</span> and <span class="math inline">\(x_i \in \Sigma^n\)</span></p>
<div class="fragment">
<p><strong>Output</strong>:</p>
<p>for each column <span class="math inline">\(k\)</span> in <span class="math inline">\(0\leq k &lt; n\)</span> we represent:</p>
<ul>
<li><p>a <em>prefix</em> array <span class="math inline">\(A_k\)</span> that stores the permutation of the rows in <span class="math inline">\(X\)</span> according to the co-lex ordering of the subsequences of <span class="math inline">\(X\)</span> from <span class="math inline">\(0\)</span> to <span class="math inline">\(k\)</span>;</p></li>
<li><p><span class="math inline">\(Y_k\)</span>, the entire matrix <span class="math inline">\(X\)</span> with <span class="math inline">\(x_i\)</span> permuted according to <span class="math inline">\(A_k\)</span> (not computed)</p></li>
<li><p>an array <span class="math inline">\(y^k = Y_k[k+1]\)</span> containg the permutation of the input <span class="math inline">\(X\)</span>’s next column, according to the row permutation in <span class="math inline">\(A_k\)</span>. This is computed on-the-fly (and not stored!) reading <span class="math inline">\(X[k+1]\)</span> by the permutation stored in <span class="math inline">\(A_k\)</span>.</p></li>
<li><p>a <em>divergence</em> array <span class="math inline">\(D_k\)</span> that stores the length of the <span class="math inline">\(LCP^{-1}\)</span> between each pair <span class="math inline">\((y_i[0,k], y_{i-1}[0,k])\)</span></p></li>
</ul>
<!--The PBWT of $X$ can be efficiently stored saving only $y^k$ for each $k, 0 \leq k < n$-->
</div>
</section>
<section id="pbwt-basics" class="slide level2">
<h2>PBWT basics</h2>
<p>The catch</p>
<blockquote>
<p><span class="math inline">\(A_k\)</span> and <span class="math inline">\(D_k\)</span> can be computed by <span class="math inline">\(A_{k-1}\)</span> and <span class="math inline">\(D_{k-1}\)</span> linearly in space/time <span class="math inline">\(O(m)\)</span><span class="citation" data-cites="durbinEfficientHaplotypeMatching2014">[<a href="#/references" role="doc-biblioref" onclick="">1</a>]</span>.</p>
</blockquote>
<ul>
<li class="fragment"><p>The overall computation time of the <span class="math inline">\(\texttt{PBWT}\)</span> (composed of arrays <span class="math inline">\(A_k\)</span> and <span class="math inline">\(D_k\)</span> for each <span class="math inline">\(k\)</span> … is <span class="math inline">\(O(n\cdot m)\)</span>.</p></li>
<li class="fragment"><p>Moreover, if we keep only <span class="math inline">\(\texttt{PBWT}_{k}\)</span> to compute the next <span class="math inline">\(\texttt{PBWT}_{k+1}\)</span>, we can do so in <span class="math inline">\(O(m)\)</span> space, performing some other computation on each <span class="math inline">\(\texttt{PBWT}\)</span> before it is discarded.</p></li>
</ul>
</section>
<section id="what-pbwt-is-mainly-used-for" class="slide level2 smaller">
<h2>What PBWT is (mainly) used for:</h2>
<ol type="1">
<li><p>computing Matching Statistics (MS) and Set Maximal Exact Matches (SMEMs)</p></li>
<li><p>computing internal matches longer than <span class="math inline">\(L\)</span> and Maximal Perfect Haplotype Blocks (MPHB)</p></li>
<li><p>Compression</p></li>
</ol>
<p>(1 and 2) are used to make statitstics on Identity-by-Descendant (IDB) segments or to identify positive selection areas (3 ) an example of pbwt used to compress VCFs <span class="citation" data-cites="Wertenbroek_2022">[<a href="#/references" role="doc-biblioref" onclick="">2</a>]</span>.</p>
</section>
<section id="computing-ms-and-smems" class="slide level2 scrollable">
<h2>Computing MS and SMEMs</h2>
<ul>
<li><p><strong>Goal</strong>: match a pattern <span class="math inline">\(P\)</span> (shorter or equal the matrix width <span class="math inline">\(n\)</span>) against <span class="math inline">\(X\)</span>.</p>
<p>We want to compute the best matching haplotype (in colex oreder) in <span class="math inline">\(X\)</span> to each subsequence of <span class="math inline">\(P\)</span> from <span class="math inline">\(0\)</span> to <span class="math inline">\(k\)</span>, for each <span class="math inline">\(k\)</span> in <span class="math inline">\(0 \leq k &lt; |P|\)</span></p></li>
<li><p>Can be computed sweeping trough the <span class="math inline">\(D\)</span> array and tracking the <em>matching statistics</em></p></li>
<li><p><strong>tools</strong>:</p>
<ul>
<li><p>mc-PBWT: multi column (parallel?) scanning of the PBWT (double and triple) <span class="citation" data-cites="shakyaMcPBWTSpaceEfficientMulticolumn2022">[<a href="#/references" role="doc-biblioref" onclick="">3</a>]</span> used for “Matches of exact Lenght <span class="math inline">\(L\)</span></p>
<hr></li>
<li><p>parallel PBWT (VCF/BCF): first version, vertically parallel + reconcile <span class="citation" data-cites="wertenbroekExploitingParallelizationPositional2023">[<a href="#/references" role="doc-biblioref" onclick="">4</a>]</span> Long Matches and SMEMs</p></li>
<li><p>HB-parallel PBWT (VCF?): last version, horizontally parallel <span class="citation" data-cites="tangHaplotypebasedParallelPBWT2025">[<a href="#/references" role="doc-biblioref" onclick="">5</a>]</span> Long Matches and SMEMs</p>
<hr></li>
<li><p>mu-pbwt: <span class="citation" data-cites="cozziPBWTLightweightRindexing2023">[<a href="#/references" role="doc-biblioref" onclick="">6</a>]</span> Run-Lenght Encoded PBWT + MS and SMEMs against Query</p></li>
<li><p>syllable-PBWT: <span class="citation" data-cites="wangSyllablePBWTSpaceefficientHaplotype2023">[<a href="#/references" role="doc-biblioref" onclick="">7</a>]</span> Syllable-packed PBWT, hexadecimal every 4 bits.. Fanno solo Long Matches (query). Non fanno SMEMs, non fanno Blocchi. Handles only VCFs and not BCFs (comparazione con Cozzi’s mu-pbwt).</p></li>
<li><p>d-pbwt:<span class="citation" data-cites="sanaullahDPBWTDynamicPositional2021">[<a href="#/references" role="doc-biblioref" onclick="">8</a>]</span>, Query of SMEMs, Long matches, All vs All</p></li>
<li><p>mu+d-pbwt: <span class="citation" data-cites="shakyaDynamicMPBWTDynamic2025">[<a href="#/references" role="doc-biblioref" onclick="">9</a>]</span>, RLE + Dynamic PBWT</p></li>
<li><p>efficient matching: <span class="citation" data-cites="naseriEfficientHaplotypeMatching2019">[<a href="#/references" role="doc-biblioref" onclick="">10</a>]</span></p></li>
<li><p>templated-PBWT <span class="citation" data-cites="freymanFastRobustIdentitybyDescent2021">[<a href="#/references" role="doc-biblioref" onclick="">11</a>]</span></p></li>
</ul></li>
</ul>
</section>
<section id="c.-internal-matches-and-blocks" class="slide level2">
<h2>C. Internal Matches and Blocks</h2>
<ul>
<li>MPHB
<ul>
<li>(Trie) <span class="citation" data-cites="cunhaIdentifyingMaximalPerfect2018">[Cunha et al, <a href="#/references" role="doc-biblioref" onclick="">12</a>]</span>, (mPBWT) <span class="citation" data-cites="alankoFindingAllMaximal2020">[Alanko et al., <a href="#/references" role="doc-biblioref" onclick="">13</a>]</span></li>
<li>(bi-directional PBWT) <span class="citation" data-cites="naseriEfficientHaplotypeBlock2021">[Naseri et al, <a href="#/references" role="doc-biblioref" onclick="">14</a>]</span> “Blocks of at least width <span class="math inline">\(W\)</span> + report exact length”</li>
</ul></li>
<li>MPHBw &amp; MHBw
<ul>
<li>(Trie) <span class="citation" data-cites="williamsMaximalPerfectHaplotype2020">[Williams et al., <a href="#/references" role="doc-biblioref" onclick="">15</a>]</span>, (mPBWT) <span class="citation" data-cites="bonizzoniMultiallelicMaximalPerfect2023">[Bonizzoni et al., <a href="#/references" role="doc-biblioref" onclick="">16</a>]</span></li>
</ul></li>
</ul>
</section>
<section id="compression" class="slide level2">
<h2>Compression</h2>
<ul>
<li>Run-Length encoding</li>
</ul>
</section>
<section id="list" class="slide level2 scrollable">
<h2>List</h2>
<p>Durbin’s (Set maximal exact matches)</p>
<ul>
<li>pbwt</li>
<li>graph-pbwt</li>
<li>mu-pbwt</li>
<li>d-pbwt</li>
<li>mc-pbwt</li>
<li>haplotype-based parallel pbwt (Naseri, zing zang sanahulla)</li>
<li>syllable-pbwt</li>
<li>tree consistent pbwt (tc-pbwt)</li>
<li>fast recomb</li>
<li>Data Structures for SMEM-Finding in the PBWT</li>
<li>Compressed Data Structures for Population-Scale Positional Burrows–Wheeler Transforms</li>
<li>multiallelic pbwt (zing, zang, sanahulla)</li>
<li>templated PBWT</li>
<li>FiMap PBWT</li>
</ul>
<p>Compression - Durbin - CoMSA (gPBWT) - GTShark</p>
<p>Haplotype blocks - bi-pbwt - haploblocks (pbwt linear)</p>
<p>Haplotype blocks with wildcards - psmoother (pbwt, c++) - wild-pbwt (pbwt, c++) - wildhap (trie, java)</p>
</section>
<section id="references" class="slide level2 smaller scrollable">
<h2>References</h2>
<div id="refs" class="references csl-bib-body" role="list">
<div id="ref-durbinEfficientHaplotypeMatching2014" class="csl-entry" role="listitem">
<div class="csl-left-margin">1. </div><div class="csl-right-inline">Durbin R (2014) Efficient haplotype matching and storage using the positional <span>Burrows</span>–<span>Wheeler</span> transform (<span>PBWT</span>). Bioinformatics 30(9):1266–1272. <a href="https://doi.org/10.1093/bioinformatics/btu014">https://doi.org/10.1093/bioinformatics/btu014</a></div>
</div>
<div id="ref-Wertenbroek_2022" class="csl-entry" role="listitem">
<div class="csl-left-margin">2. </div><div class="csl-right-inline">Wertenbroek R, Wertenbroek R, Rubinacci S, et al (2022) <span>XSI</span> - <span>A</span> genotype compression tool for compressive genomics in large biobanks. Bioinformatics (Oxford, England). <a href="https://doi.org/10.1093/bioinformatics/btac413">https://doi.org/10.1093/bioinformatics/btac413</a></div>
</div>
<div id="ref-shakyaMcPBWTSpaceEfficientMulticolumn2022" class="csl-entry" role="listitem">
<div class="csl-left-margin">3. </div><div class="csl-right-inline">Shakya P, Naseri A, Zhi D, Zhang S (2022) <a href="https://doi.org/10.1007/978-3-031-17531-2_10"><span class="nocase">mcPBWT</span>: <span class="nocase">Space-Efficient Multi-column PBWT Scanning Algorithm</span> for&nbsp;<span>Composite Haplotype Matching</span></a>. In: Bansal MS, Măndoiu I, Moussa M, et al (eds) Computational <span>Advances</span> in <span>Bio</span> and <span>Medical Sciences</span>. Springer International Publishing, Cham, pp 115–130</div>
</div>
<div id="ref-wertenbroekExploitingParallelizationPositional2023" class="csl-entry" role="listitem">
<div class="csl-left-margin">4. </div><div class="csl-right-inline">Wertenbroek R, Xenarios I, Thoma Y, Delaneau O (2023) Exploiting parallelization in positional <span>Burrows</span>–<span>Wheeler</span> transform (<span>PBWT</span>) algorithms for efficient haplotype matching and compression. Bioinformatics Advances 3(1):vbad021. <a href="https://doi.org/10.1093/bioadv/vbad021">https://doi.org/10.1093/bioadv/vbad021</a></div>
</div>
<div id="ref-tangHaplotypebasedParallelPBWT2025" class="csl-entry" role="listitem">
<div class="csl-left-margin">5. </div><div class="csl-right-inline">Tang K, Sanaullah A, Zhi D, Zhang S (2025) <a href="https://doi.org/10.1101/2025.02.04.636317">Haplotype-based <span>Parallel PBWT</span> for <span>Biobank Scale Data</span></a>. 2025.02.04.636317</div>
</div>
<div id="ref-cozziPBWTLightweightRindexing2023" class="csl-entry" role="listitem">
<div class="csl-left-margin">6. </div><div class="csl-right-inline">Cozzi D, Rossi M, Rubinacci S, et al (2023) <span><span class="math inline">\(\mu\)</span></span>- <span>PBWT</span>: A lightweight r-indexing of the <span>PBWT</span> for storing and querying <span>UK Biobank</span> data. Bioinformatics 39(9):btad552. <a href="https://doi.org/10.1093/bioinformatics/btad552">https://doi.org/10.1093/bioinformatics/btad552</a></div>
</div>
<div id="ref-wangSyllablePBWTSpaceefficientHaplotype2023" class="csl-entry" role="listitem">
<div class="csl-left-margin">7. </div><div class="csl-right-inline">Wang V, Naseri A, Zhang S, Zhi D (2023) Syllable-<span>PBWT</span> for space-efficient haplotype long-match query. Bioinformatics 39(1):btac734. <a href="https://doi.org/10.1093/bioinformatics/btac734">https://doi.org/10.1093/bioinformatics/btac734</a></div>
</div>
<div id="ref-sanaullahDPBWTDynamicPositional2021" class="csl-entry" role="listitem">
<div class="csl-left-margin">8. </div><div class="csl-right-inline">Sanaullah A, Zhi D, Zhang S (2021) D-<span>PBWT</span>: Dynamic positional <span>Burrows</span>–<span>Wheeler</span> transform. Bioinformatics 37(16):2390–2397. <a href="https://doi.org/10.1093/bioinformatics/btab117">https://doi.org/10.1093/bioinformatics/btab117</a></div>
</div>
<div id="ref-shakyaDynamicMPBWTDynamic2025" class="csl-entry" role="listitem">
<div class="csl-left-margin">9. </div><div class="csl-right-inline">Shakya P, Sanaullah A, Zhi D, Zhang S (2025) Dynamic <span><span class="math inline">\(\mu\)</span></span>-<span>PBWT</span>: <span class="nocase">Dynamic Run-length Compressed PBWT</span> for <span>Biobank Scale Data</span>. bioRxiv 2025.02.04.636479. <a href="https://doi.org/10.1101/2025.02.04.636479">https://doi.org/10.1101/2025.02.04.636479</a></div>
</div>
<div id="ref-naseriEfficientHaplotypeMatching2019" class="csl-entry" role="listitem">
<div class="csl-left-margin">10. </div><div class="csl-right-inline">Naseri A, Holzhauser E, Zhi D, Zhang S (2019) Efficient haplotype matching between a query and a panel for genealogical search. Bioinformatics 35(14):i233–i241. <a href="https://doi.org/10.1093/bioinformatics/btz347">https://doi.org/10.1093/bioinformatics/btz347</a></div>
</div>
<div id="ref-freymanFastRobustIdentitybyDescent2021" class="csl-entry" role="listitem">
<div class="csl-left-margin">11. </div><div class="csl-right-inline">Freyman WA, McManus KF, Shringarpure SS, et al (2021) Fast and <span class="nocase">Robust Identity-by-Descent Inference</span> with the <span>Templated Positional Burrows</span>–<span>Wheeler Transform</span>. Molecular Biology and Evolution 38(5):2131–2151. <a href="https://doi.org/10.1093/molbev/msaa328">https://doi.org/10.1093/molbev/msaa328</a></div>
</div>
<div id="ref-cunhaIdentifyingMaximalPerfect2018" class="csl-entry" role="listitem">
<div class="csl-left-margin">12. </div><div class="csl-right-inline">Cunha L, Diekmann Y, Kowada L, Stoye J (2018) <a href="https://doi.org/10.1007/978-3-030-01722-4_3">Identifying <span>Maximal Perfect Haplotype Blocks</span></a>. In: Alves R (ed) Advances in <span>Bioinformatics</span> and <span>Computational Biology</span>. Springer International Publishing, Cham, pp 26–37</div>
</div>
<div id="ref-alankoFindingAllMaximal2020" class="csl-entry" role="listitem">
<div class="csl-left-margin">13. </div><div class="csl-right-inline">Alanko J, Bannai H, Cazaux B, Peterlongo P, Stoye J (2020) Finding all maximal perfect haplotype blocks in linear time. Algorithms for Molecular Biology 15(1):2. <a href="https://doi.org/10.1186/s13015-020-0163-6">https://doi.org/10.1186/s13015-020-0163-6</a></div>
</div>
<div id="ref-naseriEfficientHaplotypeBlock2021" class="csl-entry" role="listitem">
<div class="csl-left-margin">14. </div><div class="csl-right-inline">Naseri A, Yue W, Zhang S, Zhi D (2021) <a href="https://doi.org/10.4230/LIPIcs.WABI.2021.19">Efficient <span>Haplotype Block Matching</span> in <span>Bi-Directional PBWT</span></a>. In: Carbone A, El-Kebir M (eds) 21st <span>International Workshop</span> on <span>Algorithms</span> in <span>Bioinformatics</span> (<span>WABI</span> 2021). Schloss Dagstuhl – Leibniz-Zentrum f<span>ü</span>r Informatik, Dagstuhl, Germany, pp 19:1–19:13</div>
</div>
<div id="ref-williamsMaximalPerfectHaplotype2020" class="csl-entry" role="listitem">
<div class="csl-left-margin">15. </div><div class="csl-right-inline">Williams L, Mumey B (2020) Maximal <span>Perfect Haplotype Blocks</span> with <span>Wildcards</span></div>
</div>
<div id="ref-bonizzoniMultiallelicMaximalPerfect2023" class="csl-entry" role="listitem">
<div class="csl-left-margin">16. </div><div class="csl-right-inline">Bonizzoni P, Della Vedova G, Pirola Y, Rizzi R, Sgrò M (2023) <a href="https://doi.org/10.1007/978-3-031-34953-9_5">Multiallelic <span>Maximal Perfect Haplotype Blocks</span> with <span>Wildcards</span> via <span>PBWT</span></a>. In: Rojas I, Valenzuela O, Rojas Ruiz F, Herrera LJ, Ortuño F (eds) Bioinformatics and <span>Biomedical Engineering</span>. Springer Nature Switzerland, Cham, pp 62–76</div>
</div>
</div>

</section></section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">

</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="index_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="index_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="index_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="index_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="index_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="index_files/libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="index_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="index_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="index_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="index_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="index_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': true,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"8\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":false},
'smaller': true,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    <script id="quarto-html-after-body" type="application/javascript">
      window.document.addEventListener("DOMContentLoaded", function (event) {
        const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
        tabsets.forEach(function(tabset) {
          const tabby = new Tabby('#' + tabset.id);
        });
        const isCodeAnnotation = (el) => {
          for (const clz of el.classList) {
            if (clz.startsWith('code-annotation-')) {                     
              return true;
            }
          }
          return false;
        }
        const onCopySuccess = function(e) {
          // button target
          const button = e.trigger;
          // don't keep focus
          button.blur();
          // flash "checked"
          button.classList.add('code-copy-button-checked');
          var currentTitle = button.getAttribute("title");
          button.setAttribute("title", "Copied!");
          let tooltip;
          if (window.bootstrap) {
            button.setAttribute("data-bs-toggle", "tooltip");
            button.setAttribute("data-bs-placement", "left");
            button.setAttribute("data-bs-title", "Copied!");
            tooltip = new bootstrap.Tooltip(button, 
              { trigger: "manual", 
                customClass: "code-copy-button-tooltip",
                offset: [0, -8]});
            tooltip.show();    
          }
          setTimeout(function() {
            if (tooltip) {
              tooltip.hide();
              button.removeAttribute("data-bs-title");
              button.removeAttribute("data-bs-toggle");
              button.removeAttribute("data-bs-placement");
            }
            button.setAttribute("title", currentTitle);
            button.classList.remove('code-copy-button-checked');
          }, 1000);
          // clear code selection
          e.clearSelection();
        }
        const getTextToCopy = function(trigger) {
            const codeEl = trigger.previousElementSibling.cloneNode(true);
            for (const childEl of codeEl.children) {
              if (isCodeAnnotation(childEl)) {
                childEl.remove();
              }
            }
            return codeEl.innerText;
        }
        const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
          text: getTextToCopy
        });
        clipboard.on('success', onCopySuccess);
        if (window.document.getElementById('quarto-embedded-source-code-modal')) {
          const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
            text: getTextToCopy,
            container: window.document.getElementById('quarto-embedded-source-code-modal')
          });
          clipboardModal.on('success', onCopySuccess);
        }
          var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
          var mailtoRegex = new RegExp(/^mailto:/);
            var filterRegex = new RegExp('/' + window.location.host + '/');
          var isInternal = (href) => {
              return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
          }
          // Inspect non-navigation links and adorn them if external
         var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
          for (var i=0; i<links.length; i++) {
            const link = links[i];
            if (!isInternal(link.href)) {
              // undo the damage that might have been done by quarto-nav.js in the case of
              // links that we want to consider external
              if (link.dataset.originalHref !== undefined) {
                link.href = link.dataset.originalHref;
              }
            }
          }
        function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
          const config = {
            allowHTML: true,
            maxWidth: 500,
            delay: 100,
            arrow: false,
            appendTo: function(el) {
                return el.closest('section.slide') || el.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'light-border',
            placement: 'bottom-start',
          };
          if (contentFn) {
            config.content = contentFn;
          }
          if (onTriggerFn) {
            config.onTrigger = onTriggerFn;
          }
          if (onUntriggerFn) {
            config.onUntrigger = onUntriggerFn;
          }
            config['offset'] = [0,0];
            config['maxWidth'] = 700;
          window.tippy(el, config); 
        }
        const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
        for (var i=0; i<noterefs.length; i++) {
          const ref = noterefs[i];
          tippyHover(ref, function() {
            // use id or data attribute instead here
            let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
            try { href = new URL(href).hash; } catch {}
            const id = href.replace(/^#\/?/, "");
            const note = window.document.getElementById(id);
            if (note) {
              return note.innerHTML;
            } else {
              return "";
            }
          });
        }
        const findCites = (el) => {
          const parentEl = el.parentElement;
          if (parentEl) {
            const cites = parentEl.dataset.cites;
            if (cites) {
              return {
                el,
                cites: cites.split(' ')
              };
            } else {
              return findCites(el.parentElement)
            }
          } else {
            return undefined;
          }
        };
        var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
        for (var i=0; i<bibliorefs.length; i++) {
          const ref = bibliorefs[i];
          const citeInfo = findCites(ref);
          if (citeInfo) {
            tippyHover(citeInfo.el, function() {
              var popup = window.document.createElement('div');
              citeInfo.cites.forEach(function(cite) {
                var citeDiv = window.document.createElement('div');
                citeDiv.classList.add('hanging-indent');
                citeDiv.classList.add('csl-entry');
                var biblioDiv = window.document.getElementById('ref-' + cite);
                if (biblioDiv) {
                  citeDiv.innerHTML = biblioDiv.innerHTML;
                }
                popup.appendChild(citeDiv);
              });
              return popup.innerHTML;
            });
          }
        }
      });
      </script>
    

</body></html>