<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.22">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Mattia S.">

<title>on the positional BWT</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="index_files/libs/clipboard/clipboard.min.js"></script>
<script src="index_files/libs/quarto-html/quarto.js" type="module"></script>
<script src="index_files/libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="index_files/libs/quarto-html/popper.min.js"></script>
<script src="index_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="index_files/libs/quarto-html/anchor.min.js"></script>
<link href="index_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="index_files/libs/quarto-html/quarto-syntax-highlighting-7b4406b7675125bc2ba204020e191172.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="index_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="index_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="index_files/libs/bootstrap/bootstrap-be9206a58a27969b3ea3001da8f26fef.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="fullcontent quarto-light">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">on the positional BWT</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Mattia S. </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<section id="bf-pbwt-results" class="level1">
<h1>bf-pbwt results</h1>
<table class="caption-top table">
<thead>
<tr class="header">
<th>BLIM timing(s)</th>
<th>chr10(18gb)</th>
<th>chr21(4.7gb)</th>
<th>chr22(4.8)</th>
<th></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>2bfpbwt-bm on master</td>
<td>690.53</td>
<td>110.47</td>
<td>116.99</td>
<td></td>
</tr>
<tr class="even">
<td>2bfpbwt on (71a045a)</td>
<td>294.98</td>
<td>74.48</td>
<td>75.54</td>
<td></td>
</tr>
<tr class="odd">
<td>2bfpbwt-bm on (hotfix)</td>
<td>753.71</td>
<td>108.56</td>
<td>113.40</td>
<td></td>
</tr>
</tbody>
</table>
<table class="caption-top table">
<thead>
<tr class="header">
<th>BLIM ram (GiB)</th>
<th>chr10(18gb)</th>
<th>chr21(4.7gb)</th>
<th>chr22(4.8)</th>
<th></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>2bfpbwt-bm on master</td>
<td>250</td>
<td>82</td>
<td>84</td>
<td></td>
</tr>
<tr class="even">
<td>2bfpbwt on (71a045a)</td>
<td>158</td>
<td>43</td>
<td>44</td>
<td></td>
</tr>
<tr class="odd">
<td>2bfpbwt-bm on (hotfix)</td>
<td>244</td>
<td>82</td>
<td>84</td>
<td></td>
</tr>
</tbody>
</table>
</section>
<section id="review" class="level1">
<h1>REVIEW</h1>
<section id="durbins-original-pbwt-httpsgithub.comricharddurbinpbwt-durbinefficienthaplotypematching2014" class="level2">
<h2 class="anchored" data-anchor-id="durbins-original-pbwt-httpsgithub.comricharddurbinpbwt-durbinefficienthaplotypematching2014">Durbin’s Original PBWT <a href="https://github.com/richarddurbin/pbwt">https://github.com/richarddurbin/pbwt, <span class="citation" data-cites="durbinEfficientHaplotypeMatching2014">[<span>1</span>]</span></a></h2>
<p><strong>Input</strong>:</p>
<p>a binary matrix <span class="math inline">\(X\)</span>, of <span class="math inline">\(m\)</span> rows and <span class="math inline">\(n\)</span> columns. <span class="math inline">\((\Sigma = \{0, 1\})\)</span>, where <span class="math inline">\(x_i\)</span> are haplotype sequences; <span class="math inline">\(0\leq i &lt; m\)</span> and <span class="math inline">\(x_i \in \Sigma^n\)</span></p>
<p><strong>Output</strong>:</p>
<p>for each column <span class="math inline">\(k\)</span> in <span class="math inline">\(0\leq k &lt; n\)</span> we represent:</p>
<ul>
<li><p>a <em>prefix</em> array <span class="math inline">\(A_k\)</span> that stores the permutation of the rows in <span class="math inline">\(X\)</span> according to the co-lex ordering of the subsequences of <span class="math inline">\(X\)</span> from <span class="math inline">\(0\)</span> to <span class="math inline">\(k\)</span>;</p></li>
<li><p><span class="math inline">\(Y_k\)</span>, the entire matrix <span class="math inline">\(X\)</span> with <span class="math inline">\(x_i\)</span> permuted according to <span class="math inline">\(A_k\)</span> (not computed)</p></li>
<li><p>an array <span class="math inline">\(y^k = Y_k[k+1]\)</span> containg the permutation of the input <span class="math inline">\(X\)</span>’s next column, according to the row permutation in <span class="math inline">\(A_k\)</span>. This is computed on-the-fly (and not stored!) reading <span class="math inline">\(X[k+1]\)</span> by the permutation stored in <span class="math inline">\(A_k\)</span>.</p></li>
<li><p>a <em>divergence</em> array <span class="math inline">\(D_k\)</span> that stores the length of the <span class="math inline">\(LCP^{-1}\)</span> between each pair <span class="math inline">\((y_i[0,k], y_{i-1}[0,k])\)</span></p></li>
</ul>
<p><strong>The catch</strong></p>
<blockquote class="blockquote">
<p><span class="math inline">\(A_k\)</span> and <span class="math inline">\(D_k\)</span> can be computed by <span class="math inline">\(A_{k-1}\)</span> and <span class="math inline">\(D_{k-1}\)</span> linearly in space/time <span class="math inline">\(O(m)\)</span><span class="citation" data-cites="durbinEfficientHaplotypeMatching2014">[<a href="#ref-durbinEfficientHaplotypeMatching2014" role="doc-biblioref">1</a>]</span>.</p>
</blockquote>
<ul>
<li><p>The overall computation time of the <span class="math inline">\(\texttt{PBWT}\)</span> (composed of arrays <span class="math inline">\(A_k\)</span> and <span class="math inline">\(D_k\)</span> for each <span class="math inline">\(k\)</span> … is <span class="math inline">\(O(n\cdot m)\)</span>.</p></li>
<li><p>Moreover, if we keep only <span class="math inline">\(\texttt{PBWT}_{k}\)</span> to compute the next <span class="math inline">\(\texttt{PBWT}_{k+1}\)</span>, we can do so in <span class="math inline">\(O(m)\)</span> space, performing some other computation on each <span class="math inline">\(\texttt{PBWT}\)</span> before it is discarded.</p></li>
</ul>
</section>
<section id="pbwt-use" class="level2">
<h2 class="anchored" data-anchor-id="pbwt-use">PBWT Use:</h2>
<ol type="1">
<li>Indexing + Querying of <em>Set Maximal Exact Matches</em> and/or <em>L-long matches</em>, computing so-called <em>matching statistics</em>
<ul>
<li>SMEMs: match a pattern <span class="math inline">\(P\)</span> against <span class="math inline">\(X\)</span>, retrive the longest matches of <span class="math inline">\(P[0,k]\)</span> subsequences in co-lex ordering.</li>
</ul></li>
<li>Computing internal all-vs-all matches longer than <span class="math inline">\(L\)</span> and/or Maximal Perfect Haplotype Blocks (MPHB)</li>
</ol>
<section id="tools" class="level4">
<h4 class="anchored" data-anchor-id="tools">Tools</h4>
<ul>
<li><p>Durbin’s pbwt <span class="citation" data-cites="durbinEfficientHaplotypeMatching2014">[<a href="#ref-durbinEfficientHaplotypeMatching2014" role="doc-biblioref">1</a>]</span> <a href="https://github.com/richarddurbin/pbwt">git</a></p>
<ul>
<li>First Definition of PBWT</li>
<li>Reads VCF or .macs simulated data</li>
<li>On-line computation</li>
<li>Computes internal <strong>L-long</strong> matches, internal <strong>SMEMs</strong>, external (query) <strong>SMEMs</strong>.</li>
</ul></li>
<li><p>mu-pbwt <span class="citation" data-cites="cozziPBWTLightweightRindexing2023">[<a href="#ref-cozziPBWTLightweightRindexing2023" role="doc-biblioref">2</a>]</span> <a href="https://github.com/dlcgold/muPBWT">git</a></p>
<ul>
<li>Run-Length encoding of the PBWT</li>
<li>reads VCF/BCF</li>
<li>indexes and queries of a pattern against the VCF</li>
<li>computes external <strong>SMEMs</strong> and <strong>L-long matches</strong></li>
</ul></li>
<li><p>d-pbwt <span class="citation" data-cites="sanaullahDPBWTDynamicPositional2021">[<a href="#ref-sanaullahDPBWTDynamicPositional2021" role="doc-biblioref">3</a>]</span> <a href="http://genome.ucf.edu/d-PBWT/">git</a></p>
<ul>
<li>Dynamic pbwt to dynamically insert/delete reads</li>
<li>reads VCF</li>
<li>indexes and queries of patterns against Panel,</li>
<li>computes external <strong>L-long matches</strong></li>
</ul></li>
<li><p>dynamic-mu-pbwt <span class="citation" data-cites="shakyaDynamicMPBWTDynamic2025">[<a href="#ref-shakyaDynamicMPBWTDynamic2025" role="doc-biblioref">4</a>]</span> <a href="https://github.com/ucfcbb/Dynamic-mu-PBWT">git</a></p>
<ul>
<li>dynamic + mu PBWT</li>
<li>reads VCF</li>
<li>indexes and queries of patterns against Panel,</li>
<li>computes external <strong>SMEMs</strong> and <strong>L-long matches</strong></li>
</ul></li>
<li><p>PBWT-Query <span class="citation" data-cites="naseriEfficientHaplotypeMatching2019">[<a href="#ref-naseriEfficientHaplotypeMatching2019" role="doc-biblioref">5</a>]</span> <a href="genome.ucf.edu/pbwt-query">git</a></p>
<ul>
<li>Code is unaccesible, may read VCF</li>
<li>Uses “Leap Arrays” to jump from differente <span class="math inline">\(D\)</span> values in <span class="math inline">\(y^k\)</span></li>
<li>Notes: several version of the algorithms are proposed
<ul>
<li><strong>PBWT-Query</strong> memory extensive, panel in memory</li>
<li><strong>L-PBWT-Query</strong> memory mapped, panel lazy-loaded; uses Leap arrays</li>
<li><strong>L-PBWT-Query</strong> memory extensive, panel in memory; uses Leap arrays</li>
</ul></li>
<li>computes external <strong>L-long matches</strong></li>
</ul></li>
<li><p>syllable-PBWT <span class="citation" data-cites="wangSyllablePBWTSpaceefficientHaplotype2023">[<a href="#ref-wangSyllablePBWTSpaceefficientHaplotype2023" role="doc-biblioref">6</a>]</span> <a href="https://github.com/ZhiGroup/Syllable-PBWT">git</a></p>
<ul>
<li>syllable-encoding of the PBWT, 64 or 128 bits</li>
<li>reads VCF</li>
<li>indexes and queries of patterns against Panel, parametrized by <span class="math inline">\(B\)</span> syllable length (<span class="math inline">\(B=\{64, 128\}\)</span>)</li>
<li>computes external <strong>L-long matches</strong> with <span class="math inline">\(L \geq 2B\)</span></li>
<li>notes: in the paper section (2.1) and Table 1, they write about “bit-PBWT” referring to the Durbin’s original PBWT used in PBWT-Query <span class="citation" data-cites="naseriEfficientHaplotypeMatching2019">[<a href="#ref-naseriEfficientHaplotypeMatching2019" role="doc-biblioref">5</a>]</span> and d-PBWT <span class="citation" data-cites="sanaullahDPBWTDynamicPositional2021">[<a href="#ref-sanaullahDPBWTDynamicPositional2021" role="doc-biblioref">3</a>]</span></li>
</ul></li>
<li><p>mc-PBWT: multi column (parallel?) scanning of the PBWT (double and triple) <span class="citation" data-cites="shakyaMcPBWTSpaceEfficientMulticolumn2022">[<a href="#ref-shakyaMcPBWTSpaceEfficientMulticolumn2022" role="doc-biblioref">7</a>]</span> used for “Matches of exact Lenght <span class="math inline">\(L\)</span></p></li>
<li><p>parallel PBWT (VCF/BCF): first version, vertically parallel + reconcile <span class="citation" data-cites="wertenbroekExploitingParallelizationPositional2023">[<a href="#ref-wertenbroekExploitingParallelizationPositional2023" role="doc-biblioref">8</a>]</span> Long Matches and SMEMs</p></li>
<li><p>HB-parallel PBWT (VCF?): last version, horizontally parallel <span class="citation" data-cites="tangHaplotypebasedParallelPBWT2025">[<a href="#ref-tangHaplotypebasedParallelPBWT2025" role="doc-biblioref">9</a>]</span> Long Matches and SMEMs</p></li>
<li><p>Trie-based Block Matching <span class="citation" data-cites="cunhaIdentifyingMaximalPerfect2018">[<a href="#ref-cunhaIdentifyingMaximalPerfect2018" role="doc-biblioref">10</a>]</span></p></li>
<li><p>Haploblocks (PBWT-based block Matching) <span class="citation" data-cites="alankoFindingAllMaximal2020">[<a href="#ref-alankoFindingAllMaximal2020" role="doc-biblioref">11</a>]</span> <a href="https://gitlab.com/bacazaux/haploblocks">git</a></p></li>
<li><p>Bi-directional PBWT <span class="citation" data-cites="naseriEfficientHaplotypeBlock2021">[<a href="#ref-naseriEfficientHaplotypeBlock2021" role="doc-biblioref">12</a>]</span> <a href="https://github.com/ZhiGroup/bi-PBWT">git</a></p></li>
<li><p>WildHap (Trie-based * block matching) <span class="citation" data-cites="williamsMaximalPerfectHaplotype2020">[<a href="#ref-williamsMaximalPerfectHaplotype2020" role="doc-biblioref">13</a>]</span> <a href="williamsExtendingMaximalPerfect2020">git</a></p></li>
<li><p>Wild-pBWT (PBWT-based * block matching) <span class="citation" data-cites="bonizzoniMultiallelicMaximalPerfect2023">[<a href="#ref-bonizzoniMultiallelicMaximalPerfect2023" role="doc-biblioref">14</a>]</span> <a href="https://github.com/AlgoLab/Wild-pBWT">git</a></p></li>
<li><p>templated PBWT <span class="citation" data-cites="freymanFastRobustIdentitybyDescent2021">[<a href="#ref-freymanFastRobustIdentitybyDescent2021" role="doc-biblioref">15</a>]</span> <a href="https://github.com/23andMe/phasedibd">git</a></p></li>
<li><p>psmoother <span class="citation" data-cites="yuePsmootherEfficientPBWT2022">[<a href="#ref-yuePsmootherEfficientPBWT2022" role="doc-biblioref">16</a>]</span> <a href="https://github.com/ZhiGroup/P-smoother">git</a></p></li>
<li><p>tree consistent pbwt (tc-pbwt)</p></li>
<li><p>fast recomb</p></li>
<li><p>Data Structures for SMEM-Finding in the PBWT</p></li>
<li><p>Compressed Data Structures for Population-Scale Positional Burrows–Wheeler Transforms</p></li>
<li><p>multiallelic pbwt (zing, zang, sanahulla)</p></li>
<li><p>templated PBWT</p></li>
<li><p>FiMap PBWT</p></li>
</ul>
<p>Compression - Durbin - CoMSA (gPBWT) - GTShark</p>
</section>
</section>
<section id="references" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="references">References</h2>
<div id="refs" class="references csl-bib-body" role="list">
<div id="ref-durbinEfficientHaplotypeMatching2014" class="csl-entry" role="listitem">
<div class="csl-left-margin">1. </div><div class="csl-right-inline">Durbin R (2014) Efficient haplotype matching and storage using the positional <span>Burrows</span>–<span>Wheeler</span> transform (<span>PBWT</span>). Bioinformatics 30(9):1266–1272. <a href="https://doi.org/10.1093/bioinformatics/btu014">https://doi.org/10.1093/bioinformatics/btu014</a></div>
</div>
<div id="ref-cozziPBWTLightweightRindexing2023" class="csl-entry" role="listitem">
<div class="csl-left-margin">2. </div><div class="csl-right-inline">Cozzi D, Rossi M, Rubinacci S, et al (2023) <span><span class="math inline">\(\mu\)</span></span>- <span>PBWT</span>: A lightweight r-indexing of the <span>PBWT</span> for storing and querying <span>UK Biobank</span> data. Bioinformatics 39(9):btad552. <a href="https://doi.org/10.1093/bioinformatics/btad552">https://doi.org/10.1093/bioinformatics/btad552</a></div>
</div>
<div id="ref-sanaullahDPBWTDynamicPositional2021" class="csl-entry" role="listitem">
<div class="csl-left-margin">3. </div><div class="csl-right-inline">Sanaullah A, Zhi D, Zhang S (2021) D-<span>PBWT</span>: Dynamic positional <span>Burrows</span>–<span>Wheeler</span> transform. Bioinformatics 37(16):2390–2397. <a href="https://doi.org/10.1093/bioinformatics/btab117">https://doi.org/10.1093/bioinformatics/btab117</a></div>
</div>
<div id="ref-shakyaDynamicMPBWTDynamic2025" class="csl-entry" role="listitem">
<div class="csl-left-margin">4. </div><div class="csl-right-inline">Shakya P, Sanaullah A, Zhi D, Zhang S (2025) Dynamic <span><span class="math inline">\(\mu\)</span></span>-<span>PBWT</span>: <span class="nocase">Dynamic Run-length Compressed PBWT</span> for <span>Biobank Scale Data</span>. bioRxiv 2025.02.04.636479. <a href="https://doi.org/10.1101/2025.02.04.636479">https://doi.org/10.1101/2025.02.04.636479</a></div>
</div>
<div id="ref-naseriEfficientHaplotypeMatching2019" class="csl-entry" role="listitem">
<div class="csl-left-margin">5. </div><div class="csl-right-inline">Naseri A, Holzhauser E, Zhi D, Zhang S (2019) Efficient haplotype matching between a query and a panel for genealogical search. Bioinformatics 35(14):i233–i241. <a href="https://doi.org/10.1093/bioinformatics/btz347">https://doi.org/10.1093/bioinformatics/btz347</a></div>
</div>
<div id="ref-wangSyllablePBWTSpaceefficientHaplotype2023" class="csl-entry" role="listitem">
<div class="csl-left-margin">6. </div><div class="csl-right-inline">Wang V, Naseri A, Zhang S, Zhi D (2023) Syllable-<span>PBWT</span> for space-efficient haplotype long-match query. Bioinformatics 39(1):btac734. <a href="https://doi.org/10.1093/bioinformatics/btac734">https://doi.org/10.1093/bioinformatics/btac734</a></div>
</div>
<div id="ref-shakyaMcPBWTSpaceEfficientMulticolumn2022" class="csl-entry" role="listitem">
<div class="csl-left-margin">7. </div><div class="csl-right-inline">Shakya P, Naseri A, Zhi D, Zhang S (2022) <a href="https://doi.org/10.1007/978-3-031-17531-2_10"><span class="nocase">mcPBWT</span>: <span class="nocase">Space-Efficient Multi-column PBWT Scanning Algorithm</span> for&nbsp;<span>Composite Haplotype Matching</span></a>. In: Bansal MS, Măndoiu I, Moussa M, et al (eds) Computational <span>Advances</span> in <span>Bio</span> and <span>Medical Sciences</span>. Springer International Publishing, Cham, pp 115–130</div>
</div>
<div id="ref-wertenbroekExploitingParallelizationPositional2023" class="csl-entry" role="listitem">
<div class="csl-left-margin">8. </div><div class="csl-right-inline">Wertenbroek R, Xenarios I, Thoma Y, Delaneau O (2023) Exploiting parallelization in positional <span>Burrows</span>–<span>Wheeler</span> transform (<span>PBWT</span>) algorithms for efficient haplotype matching and compression. Bioinformatics Advances 3(1):vbad021. <a href="https://doi.org/10.1093/bioadv/vbad021">https://doi.org/10.1093/bioadv/vbad021</a></div>
</div>
<div id="ref-tangHaplotypebasedParallelPBWT2025" class="csl-entry" role="listitem">
<div class="csl-left-margin">9. </div><div class="csl-right-inline">Tang K, Sanaullah A, Zhi D, Zhang S (2025) <a href="https://doi.org/10.1101/2025.02.04.636317">Haplotype-based <span>Parallel PBWT</span> for <span>Biobank Scale Data</span></a>. 2025.02.04.636317</div>
</div>
<div id="ref-cunhaIdentifyingMaximalPerfect2018" class="csl-entry" role="listitem">
<div class="csl-left-margin">10. </div><div class="csl-right-inline">Cunha L, Diekmann Y, Kowada L, Stoye J (2018) <a href="https://doi.org/10.1007/978-3-030-01722-4_3">Identifying <span>Maximal Perfect Haplotype Blocks</span></a>. In: Alves R (ed) Advances in <span>Bioinformatics</span> and <span>Computational Biology</span>. Springer International Publishing, Cham, pp 26–37</div>
</div>
<div id="ref-alankoFindingAllMaximal2020" class="csl-entry" role="listitem">
<div class="csl-left-margin">11. </div><div class="csl-right-inline">Alanko J, Bannai H, Cazaux B, Peterlongo P, Stoye J (2020) Finding all maximal perfect haplotype blocks in linear time. Algorithms for Molecular Biology 15(1):2. <a href="https://doi.org/10.1186/s13015-020-0163-6">https://doi.org/10.1186/s13015-020-0163-6</a></div>
</div>
<div id="ref-naseriEfficientHaplotypeBlock2021" class="csl-entry" role="listitem">
<div class="csl-left-margin">12. </div><div class="csl-right-inline">Naseri A, Yue W, Zhang S, Zhi D (2021) <a href="https://doi.org/10.4230/LIPIcs.WABI.2021.19">Efficient <span>Haplotype Block Matching</span> in <span>Bi-Directional PBWT</span></a>. In: Carbone A, El-Kebir M (eds) 21st <span>International Workshop</span> on <span>Algorithms</span> in <span>Bioinformatics</span> (<span>WABI</span> 2021). Schloss Dagstuhl – Leibniz-Zentrum f<span>ü</span>r Informatik, Dagstuhl, Germany, pp 19:1–19:13</div>
</div>
<div id="ref-williamsMaximalPerfectHaplotype2020" class="csl-entry" role="listitem">
<div class="csl-left-margin">13. </div><div class="csl-right-inline">Williams L, Mumey B (2020) Maximal <span>Perfect Haplotype Blocks</span> with <span>Wildcards</span></div>
</div>
<div id="ref-bonizzoniMultiallelicMaximalPerfect2023" class="csl-entry" role="listitem">
<div class="csl-left-margin">14. </div><div class="csl-right-inline">Bonizzoni P, Della Vedova G, Pirola Y, Rizzi R, Sgrò M (2023) <a href="https://doi.org/10.1007/978-3-031-34953-9_5">Multiallelic <span>Maximal Perfect Haplotype Blocks</span> with <span>Wildcards</span> via <span>PBWT</span></a>. In: Rojas I, Valenzuela O, Rojas Ruiz F, Herrera LJ, Ortuño F (eds) Bioinformatics and <span>Biomedical Engineering</span>. Springer Nature Switzerland, Cham, pp 62–76</div>
</div>
<div id="ref-freymanFastRobustIdentitybyDescent2021" class="csl-entry" role="listitem">
<div class="csl-left-margin">15. </div><div class="csl-right-inline">Freyman WA, McManus KF, Shringarpure SS, et al (2021) Fast and <span class="nocase">Robust Identity-by-Descent Inference</span> with the <span>Templated Positional Burrows</span>–<span>Wheeler Transform</span>. Molecular Biology and Evolution 38(5):2131–2151. <a href="https://doi.org/10.1093/molbev/msaa328">https://doi.org/10.1093/molbev/msaa328</a></div>
</div>
<div id="ref-yuePsmootherEfficientPBWT2022" class="csl-entry" role="listitem">
<div class="csl-left-margin">16. </div><div class="csl-right-inline">Yue W, Naseri A, Wang V, Shakya P, Zhang S, Zhi D (2022) P-smoother: Efficient <span>PBWT</span> smoothing of large haplotype panels. Bioinformatics Advances 2(1):vbac045. <a href="https://doi.org/10.1093/bioadv/vbac045">https://doi.org/10.1093/bioadv/vbac045</a></div>
</div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>